File: BookProvider/Java/BookProvider/app/src/main/java/com/qty/bookprovider/BookProviderActivity.java
package com.qty.bookprovider;

import androidx.appcompat.app.AppCompatActivity;

import android.os.Bundle;
import android.util.Log;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.widget.TextView;

public class BookProviderActivity extends AppCompatActivity implements IReportBack {

    private static final String TAG = "BookProvider";
    private ProviderTester providerTester;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        providerTester = new ProviderTester(this, this);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        super.onCreateOptionsMenu(menu);
        MenuInflater inflater = getMenuInflater();
        inflater.inflate(R.menu.main_menu, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        appendMenuItemText(item);
        switch (item.getItemId()) {
            case R.id.menu_add:
                providerTester.addBook();
                break;

            case R.id.menu_display_table:
                providerTester.showBooks();
                break;

            case R.id.menu_delete:
                providerTester.removeBook();
                break;

            case R.id.menu_update:
                providerTester.updateBook();
                break;

            case R.id.menu_da_clear:
                emptyText();
                break;
        }
        return true;
    }

    private TextView getTextView() {
        return (TextView) this.findViewById(R.id.text1);
    }

    protected void appendMenuItemText(MenuItem menuItem) {
        String title = menuItem.getTitle().toString();
        TextView tv = getTextView();
        tv.setText(tv.getText() + "\n" + title);
    }

    protected void emptyText() {
        TextView tv = getTextView();
        tv.setText("");
    }

    private void appendText(String s) {
        TextView tv = getTextView();
        tv.setText(tv.getText() + "\n" + s);
        Log.d(TAG, s);
    }

    public void reportBack(String tag, String message) {
        this.appendText(tag + ":" + message);
        Log.d(tag, message);
    }

    //Implement save/restore
    protected void onRestoreInstanceState(Bundle savedInstanceState) {
        super.onRestoreInstanceState(savedInstanceState);

        String st = savedInstanceState.getString("debugViewText");
        if (st == null) return;
        TextView tv = getTextView();
        tv.setText(st);
        Log.d(TAG, "Restored state");
    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);

        //save state
        TextView tv = getTextView();
        String t = tv.getText().toString();
        outState.putString("debugViewText", t);
        Log.d(TAG, "Saved state");
    }

}

File: BookProvider/Kotlin/BookProvider/app/src/main/java/com/qty/bookprovider/BookProviderActivity.kt
package com.qty.bookprovider

import android.annotation.SuppressLint
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.util.Log
import android.view.Menu
import android.view.MenuItem
import android.widget.TextView

class BookProviderActivity : AppCompatActivity(), IReportBack {

    private val providerTester = ProviderTester(this, this)

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
    }

    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
        super.onCreateOptionsMenu(menu)
        menuInflater.inflate(R.menu.main_menu, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        appendMenuItemText(item)
        when (item.itemId) {
            R.id.menu_da_clear -> emptyText()
            R.id.menu_add -> providerTester.addBook()
            R.id.menu_display_table -> providerTester.showBooks()
            R.id.menu_delete -> providerTester.removeBook()
            R.id.menu_update -> providerTester.updateBook()
        }
        return true
    }

    @SuppressLint("WrongViewCast")
    private fun getTextView(): TextView {
        return findViewById(R.id.text1)
    }

    private fun appendMenuItemText(menuItem: MenuItem) {
        val tv = getTextView()
        tv.text = String.format("%s\n%s", tv.text, menuItem.title)
    }

    private fun emptyText() {
        getTextView().text = ""
    }

    private fun appendText(s: String) {
        val tv = getTextView()
        tv.text = String.format("%s\n%s", tv.text, s)
    }

    override fun reportBack(tag: String, message: String) {
        appendText("$tag:$message")
        Log.d(TAG, message)
    }

    override fun onRestoreInstanceState(savedInstanceState: Bundle) {
        super.onRestoreInstanceState(savedInstanceState)

        val st = savedInstanceState.getString("debugViewText") ?: return
        val tv = getTextView()
        tv.text = st
        Log.d(TAG, "Restored state")
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        // save state
        outState.putString("debugViewText", getTextView().text.toString())
        Log.d(TAG, "Saved state")
    }

    companion object {
        const val TAG = "BookProvider"
    }
}